server/src/test/java/io/crate/planner/WhereClauseOptimizerTest.java

```diff
@@ -78,6 +77,26 @@ month as date_bin(INTERVAL '28' DAY, ts, 0)
                 """,
                 new PartitionName(new RelationName("doc", "partdatebin"), List.of("1676352000000")).asIndexName(),
                 new PartitionName(new RelationName("doc", "partdatebin"), List.of("1687767893000")).asIndexName()
+            )
+            // Important, ts_month has a type different from date_trunc return type to provoke implicit cast addition
+            .addPartitionedTable("""
+                create table partdatetrunc (
+                    ts TIMESTAMP WITHOUT TIME ZONE,
+                    ts_month TIMESTAMP as date_trunc('month', ts)
+                ) partitioned by (ts_month)
+                """,
+                new PartitionName(new RelationName("doc", "partdatetrunc"), List.of("1676352000000")).asIndexName(),
+                new PartitionName(new RelationName("doc", "partdatetrunc"), List.of("1687767893000")).asIndexName()
+            )
+            // Important, ts_month doesn't have type declared and expression is just a CAST.
```

Comment:
```suggestion
            // Important, ts_month doesn't have type explicitly declared but implicitly due to the CAST expression.
```

Reply:
That's not what I wanted to say. 
Basically type is inferred but not due to CAST, it can be any expression and type is inferred from expression type.
CAST here is another important "component" to make it fail without fixup.